readthedocs/buildconfig/__init__.py

```diff
@@ -0,0 +1,187 @@
+"""Validator for the RTD configuration file."""
+
+from __future__ import division, print_function, unicode_literals
+
+from os import path
+
+import six
+import yamale
+from yamale.validators import DefaultValidators, Validator
+
+V2_SCHEMA = path.join(
+    path.dirname(__file__),
+    '../rtd_tests/fixtures/spec/v2/schema.yml'
+)
+
+ALL = 'all'
+
+DEFAULT_VALUES = {
+    'formats': [],
+    'conda': None,
+    'build': {
+        'image': '2.0',
+    },
+    'python': {
+        'version': '3.6',
+        'requirements': None,
+        'install': None,
+        'extra_requirements': [],
+        'system_packages': False,
+    },
+    'sphinx': {
+        'configuration': None,
+        'fail_on_warning': False,
+    },
+    'mkdocs': {
+        'configuration': None,
+        'fail_on_warning': False,
+    },
+    'submodules': {
+        'include': [],
+        'exclude': [],
+        'recursive': False,
+    },
+    'redirects': None,
+}
+
+CONSTRAINTS = {
+    'Documentation type': {
+        'unique': ['sphinx', 'mkdocs'],
+        'default': 'sphinx',
+    },
+    'Submodules': {
+        'unique': ['submodules.include', 'submodules.exclude'],
+        'default': 'submodules.include',
+    }
+}
+
+
+def flatten(dic, keep_key=False, position=None):
+    """
+    Returns a flattened dictionary
+
+    Given a dictionary of nested dictionaries, it returns
+    a dictionary with all nested keys at the top level.
+
+    For example
+
+    {
+        'key': 'value',
+        'nested': {
+            'subkey': 'value'
+        },
+    }
+
+    Becomes
+
+    {
+        'key': value,
+        'nested': {...},
+        'nested.subkey': 'value'
+    }
+    """
+    child = {}
+
+    for key, value in dic.items():
+        if isinstance(key, six.string_types):
+            key = key.replace('.', '_')
+        if position:
+            item_position = '{}.{}'.format(position, key)
+        else:
+            item_position = key
+
+        if isinstance(value, dict):
+            child.update(
+                flatten(dic[key], keep_key, item_position)
+            )
+            if keep_key:
+                child[item_position] = value
+        else:
+            child[item_position] = value
+
+    return child
```

Comment:
It makes it easy to manipulate and check the keys, also is better to describe constraints, for example 

https://github.com/stsewd/readthedocs.org/blob/yaml-schema-yamale/readthedocs/buildconfig/__init__.py#L176-L180

We don't have to use recursion or several nested bucles to do things like that

Reply:
It's not clear if this is still required. It seems like a confusing way to do a nested dict merge. I think if we are implementing defaults through yamale there is a tighter integration rather than keeping this data in a dictionary.